### PR TITLE
move overscaled tile zoom level logic further into TileCoord

### LINKED ISSUE #1912: separate Tile's style zoom level from it's data zoom level
The current implementation of overscaled tile reparsing abuses the `z` coordinate to store the zoom level that needs to be used for style parsing. The real `z` is recovered using sourceMaxZoom. It's really hard to follow how this works or if it works.

A TileCoord should have two separate z's. One for the style and one for the actual vectortile.

### PR DESCRIPTION
**Pull request by ansis**
fixes #1912

The goal is to contain all the weird logic for representing overscaled tiles within TileCoord.

TileCoord's constructor now requires `sourceMaxZoom`, which is used to calculate the TileCoord's data zoom level and for calculating child and parent TileCoords.

Most of the test changes are just passing `w` and `sourceMaxZoom` to TileCoord's constructor. There are a couple new overscaled tile tests are added for `TileCoord.children`, `TileCoord.parent`, `TileCoord.cover`.

:eyes: @jfirebaugh @lucaswoj

### CODE REVIEW

**Comment by lucaswoj** [comment]
The `TileCoord` class is coupled so tightly to the `Tile` class that I feel we should refactor to eliminate `TileCoord` altogether by

 - moving all `TileCoord` methods to `Tile` or `TilePyramid`
 - replacing all methods that return `TileCoord`s with methods that return `Tile`s

**Comment by ansis** [comment]
> There are a couple places where we still need x, y, z tuples but that should be as lightweight and simple as possible.

TileCoord is basically a (x, y, z, w, dataZ) tuple with methods for manipulating it. I added a6bf0ad which splits `tileCover` into a separate file to make this clearer.

**Comment by lucaswoj** [comment]
I'm still mildly :-1: on this change. `sourceMaxZoom` only makes sense in the context of vector tiles. Having to pass around a separate `sourceMaxZoom` better than introducing this coupling. 

a6bf0ad748ffcc1a37030477b6ec818682fdc662 is great, either way :+1:

**Comment by ansis** [comment]
> Having to pass around a separate sourceMaxZoom better than introducing this coupling.

If we want it separate then I think we should move the parent/child tile functions into tile pyramid and make TileID even simpler. I feel like sourceMaxZoom and the functions should live in the same place.

This doesn't seem like the right solution to the problem. I'll open a separate discussion issue. I know @kkaefer has been looking at this problem in -native.
